### PR TITLE
Offline bootstrap for chocolatey

### PR DESCRIPTION
**Pull request by twangboy**
### What does this PR do?
Adds the ability to install Chocolatey by specifying a source.
Adds a dotnet salt util (`salt.utils.win_dotnet`) for working with versions. Includes the following functions:
- `dotnet.versions`
- `dotnet.versions_list`
- `dotnet.versions_details`
- `dotnet.version_atleast`

### What issues does this PR fix or reference?
https://github.com/saltstack/salt/issues/45969

### Tests written?
No

### Commits signed with GPG?
Yes

### CODE REVIEW

**Comment by garethgreenaway** [comment]
@twangboy Looks like there is a lint error.

File: doc/topics/releases/neon.rst [hunk 1]
```diff
@@ -323,13 +323,31 @@ Module Changes
 - The :py:func:`yumpkg <salt.modules.yumpkg>` module has been updated to support
   VMWare's Photon OS, which uses tdnf (a C implementation of dnf).
 
+- The :py:func:`chocolatey.bootstrap <salt.modules.chocolatey.bootstrap>` function
```
Comment (cachedout) [inline doc/topics/releases/neon.rst]: Missing period at end of sentence.

File: doc/topics/releases/neon.rst [hunk 2]
```diff
@@ -323,13 +323,31 @@ Module Changes
...
+  has been updated to support offline installation
+
+- The :py:func:`chocolatey.unbootstrap <salt.modules.chocolatey.unbootstrap>` function
+  has been added to uninstall Chocolatey
```
Comment (cachedout) [inline doc/topics/releases/neon.rst]: Also missing punctuation.

File: doc/topics/releases/neon.rst [hunk 3]
```diff
@@ -323,13 +323,31 @@ Module Changes
...
 ==============
 
 - The :py:func:`saltutil.sync_auth <salt.runners.saltutil.sync_auth>` function
   has been added to sync loadable auth modules. :py:func:`saltutil.sync_all <salt.runners.saltutil.sync_all>`
   will also include these modules.
 
+Util Changes
+============
+
+- The :py:func:`win_dotnet <salt.utils.win_dotnet>` salt util has been added to
```
Comment (cachedout) [inline doc/topics/releases/neon.rst]: s/Salt/salt/

File: salt/modules/chocolatey.py [hunk 4]
```diff
@@ -109,6 +111,10 @@ def _find_chocolatey(context, salt):
     return choc_path
 
 
+def _dotnet_versions():
```
Comment (cachedout) [inline salt/modules/chocolatey.py]: What's the purpose of this? I don't see it called anywhere.

File: salt/modules/chocolatey.py [hunk 5]
```diff
@@ -140,18 +146,47 @@ def bootstrap(force=False):
...
+        be a problem on Windows versions 2012/8 and later
+
+    Args:
+
+        force (bool):
+            Run the bootstrap process even if Chocolatey is found in the path.
+
+        source (str):
+            The location of the ``.nupkg`` file or ``.ps1`` file to run from an
+            alternate location. This can be one of the following types of urls:
```
Comment (cachedout) [inline salt/modules/chocolatey.py]: URLs

File: salt/modules/chocolatey.py [hunk 6]
```diff
@@ -161,71 +196,191 @@ def bootstrap(force=False):
...
+            # Download the KB
+            try:
+                log.debug('Downloading PowerShell...')
+                __salt__['cp.get_url'](path=url, dest=dest)
+            except MinionError:
+                err = 'Failed to download PowerShell KB for {0}' \
+                      ''.format(__grains__['osrelease'])
+                if source:
+                    err = '{0}: PowerShell is required to bootstrap ' \
+                          'Chocolatey with Source'.format(err)
```
Comment (cachedout) [inline salt/modules/chocolatey.py]: This would likely read more cleanly if you didn't reuse `err`.

File: salt/utils/win_dotnet.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        dict: A dictionary of details for each version on the system. Contains
+        the following keys:
+            - version: The version installed
+            - service_pack: The service pack for the version installed
+            - full: The full version name including the service pack
+    '''
+    return versions()['details']
+
+
+def version_atleast(version):
```
Comment (cachedout) [inline salt/utils/win_dotnet.py]: This seems like it should be `version_at_least` to be consistent with breaking between words.

**Review by cachedout — APPROVED** [review]

**Pull request merged**